website/docs/r/edgegateway_settings.html.markdown

```diff
@@ -0,0 +1,94 @@
+organization user.
+
+The main use case of this resource is to allow both providers and tenants to change edge gateways global settings (such as
+enabling load balancing or firewall) when the edge gateway was created outside of terraform.
+
+~> **Warning:** The edge gateway settings info is tied to an edge gateway. Thus, there could be only one instance per 
+edge gateway. Using a different definition for the same edge gateway ID will result in a previous instance to be overwritten.
+
+Supported in provider *v3.0+*
+
+## Example Usage
+
+```hcl
+data "vcd_edgegateway" "egw" {
+  name = "my-egw"
+}
+
+resource "vcd_edgegateway_settings" "egw-settings" {
+  edge_gateway_id         = data.vcd_edgegateway.egw.id
+  lb_enabled              = true
+  lb_acceleration_enabled = true
+  lb_logging_enabled      = true
+  lb_loglevel             = "debug"
+
+  fw_enabled                      = true
+  fw_default_rule_logging_enabled = true
+  fw_default_rule_action          = "accept"
+}
+```
+
+-> **Tip:** Although this resource changes values in the edge gateway referenced as a data source, due to how Terraform works, the state
+of the edge gateway doesn't get updated. To reconcile the state of the data source with the values modified in `vcd_edgegateway_settings`,
+you need to run `terraform refresh` after `apply`.
+
+=> **Note:** Although tenants can enable load balancing using this resource, they can't set the properties `lb_logging_enabled` and `lb_loglevel`.
+
+## Argument Reference
+
+The following arguments are supported:
+
+* `org` - (Optional) The name of organization to which the VDC belongs. Optional if defined at provider level.
+* `vdc` - (Optional) The name of VDC that owns the edge gateway. Optional if defined at provider level. 
+* `edge_gateway_name` - (Optional) A unique name for the edge gateway. (Required if `edge_gateway_id` is not set)
+* `edge_gateway_id` - (Optional) The edge gateway ID. (Required if `edge_gateway_name` is not set)
+* `lb_enabled` - (Optional) Enable load balancing. Default is `false`.
+* `lb_acceleration_enabled` - (Optional) Enable to configure the load balancer.
+* `lb_logging_enabled` - (Optional) Enables the edge gateway load balancer to collect traffic logs.
+Default is `false`. Note: **only System administrators can change this property**. It is ignored by API for Org users.
+* `lb_loglevel` - (Optional) Choose the severity of events to be logged. One of `emergency`,
+`alert`, `critical`, `error`, `warning`, `notice`, `info`, `debug`. Note: **only System administrators can change this property**. It is ignored by API for Org users.
+* `fw_enabled` (Optional) Enable firewall. Default `true`.
+* `fw_default_rule_logging_enabled` (Optional) Enable default firewall rule (last in the processing 
+order) logging. Default `false`.
+* `fw_default_rule_action` (Optional) Default firewall rule (last in the processing order) action.
+One of `accept` or `deny`. Default `deny`.
```

Comment:
For an end-user two big questions arise with current approach when you start using it:
1. Why these parameters are duplicated in `vcd_edgegateway` resources.
2. What to do with these duplications? Which resource takes the priority? What happens if both resources define same, for example `lb_enabled`, setting with different values?

Should there be a recommendation in the docs to **not** use this resource if you're a sysadmin? Or what's the best advice to the user, esp. if he's sysadmin.

Here's an illustration of what happens if I have different values between these two resources. Back and forth cycling between the `apply` runs:

```
##############################
$ terraform apply

Terraform will perform the following actions:

  # vcd_edgegateway_settings.egw-settings will be created
  + resource "vcd_edgegateway_settings" "egw-settings" {
      + edge_gateway_id                 = "urn:vcloud:gateway:68a6a97d-6e99-4d73-964a-62806b66fae0"
      + edge_gateway_name               = (known after apply)
      + fw_default_rule_action          = "accept"
      + fw_default_rule_logging_enabled = true
      + fw_enabled                      = true
      + id                              = (known after apply)
      + lb_acceleration_enabled         = true
      + lb_enabled                      = false
      + lb_logging_enabled              = true
      + lb_loglevel                     = "debug"
    }

Plan: 1 to add, 0 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

vcd_edgegateway_settings.egw-settings: Creating...
vcd_edgegateway_settings.egw-settings: Still creating... [10s elapsed]
vcd_edgegateway_settings.egw-settings: Still creating... [20s elapsed]
vcd_edgegateway_settings.egw-settings: Creation complete after 21s [id=urn:vcloud:gateway:68a6a97d-6e99-4d73-964a-62806b66fae0]

Apply complete! Resources: 1 added, 0 changed, 0 destroyed.

##############################
$ terraform apply

vcd_external_network.demo_ext_net: Refreshing state... [id=urn:vcloud:network:246abcc7-a21a-4b87-b7b3-c93e758c5b69]
vcd_org.demo_org: Refreshing state... [id=urn:vcloud:org:7b6f9c28-931f-44fd-86d2-56bf678ce922]
vcd_org_vdc.demo_vdc: Refreshing state... [id=urn:vcloud:vdc:84427075-c181-4616-b3db-4a51d346fde2]
vcd_edgegateway.demo_gw: Refreshing state... [id=urn:vcloud:gateway:68a6a97d-6e99-4d73-964a-62806b66fae0]
vcd_edgegateway_settings.egw-settings: Refreshing state... [id=urn:vcloud:gateway:68a6a97d-6e99-4d73-964a-62806b66fae0]
vcd_network_routed.demo_routed_net_web: Refreshing state... [id=urn:vcloud:network:ec7a5a64-4eb3-48a5-92c0-8bbab1c0b5da]
vcd_nsxv_firewall_rule.demo_nsxv_fw_rule: Refreshing state... [id=131078]
vcd_vapp.demo_vapp: Refreshing state... [id=urn:vcloud:vapp:591abb6c-1ce0-443d-a4b2-e897910f4f6e]
vcd_vapp_org_network.demo_vapp_net_org: Refreshing state... [id=urn:vcloud:network:82517916-8256-4d8c-9446-f7ed25da52a0]
vcd_vapp_network.demo_vapp_net_w_org: Refreshing state... [id=urn:vcloud:network:348fa5ae-1f14-4dbd-be59-1034768835ec]
vcd_vapp_firewall_rules.vapp_org_fw: Refreshing state... [id=urn:vcloud:network:82517916-8256-4d8c-9446-f7ed25da52a0]
vcd_vapp_firewall_rules.vapp_fw: Refreshing state... [id=urn:vcloud:network:348fa5ae-1f14-4dbd-be59-1034768835ec]
vcd_vapp_vm.demo_vm_wordpress: Refreshing state... [id=urn:vcloud:vm:1ea00631-9846-4c85-a411-cedc6e392512]
vcd_nsxv_dnat.demo_dnat_wordpress_https: Refreshing state... [id=196609]
vcd_vapp_nat_rules.vapp_nat: Refreshing state... [id=urn:vcloud:network:348fa5ae-1f14-4dbd-be59-1034768835ec]

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  ~ update in-place

Terraform will perform the following actions:

  # vcd_edgegateway.demo_gw will be updated in-place
  ~ resource "vcd_edgegateway" "demo_gw" {
        advanced                        = true
        configuration                   = "compact"
        default_external_network_ip     = "10.150.160.116"
        default_gateway_network         = "demo-ext-net"
        description                     = "VMworld Demo EG"
        distributed_routing             = false
        external_network_ips            = [
            "10.150.160.116",
        ]
        fips_mode_enabled               = false
      ~ fw_default_rule_action          = "accept" -> "deny"
        fw_default_rule_logging_enabled = true
        fw_enabled                      = true
        ha_enabled                      = false
        id                              = "urn:vcloud:gateway:68a6a97d-6e99-4d73-964a-62806b66fae0"
      ~ lb_acceleration_enabled         = true -> false
      ~ lb_enabled                      = false -> true
      ~ lb_logging_enabled              = true -> false
      ~ lb_loglevel                     = "debug" -> "info"
        name                            = "demo-gw"
        use_default_route_for_dns_relay = false

        external_network {
            enable_rate_limit   = false
            incoming_rate_limit = 0
            name                = "demo-ext-net"
            outgoing_rate_limit = 0

            subnet {
                gateway               = "10.150.191.253"
                netmask               = "255.255.224.0"
                use_for_default_route = true
            }
        }
    }

Plan: 0 to add, 1 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

vcd_edgegateway.demo_gw: Modifying... [id=urn:vcloud:gateway:68a6a97d-6e99-4d73-964a-62806b66fae0]
vcd_edgegateway.demo_gw: Still modifying... [id=urn:vcloud:gateway:68a6a97d-6e99-4d73-964a-62806b66fae0, 10s elapsed]
vcd_edgegateway.demo_gw: Still modifying... [id=urn:vcloud:gateway:68a6a97d-6e99-4d73-964a-62806b66fae0, 20s elapsed]
vcd_edgegateway.demo_gw: Modifications complete after 25s [id=urn:vcloud:gateway:68a6a97d-6e99-4d73-964a-62806b66fae0]

Apply complete! Resources: 0 added, 1 changed, 0 destroyed.

##############################
$ terraform apply

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  ~ update in-place

Terraform will perform the following actions:

  # vcd_edgegateway_settings.egw-settings will be updated in-place
  ~ resource "vcd_edgegateway_settings" "egw-settings" {
        edge_gateway_id                 = "urn:vcloud:gateway:68a6a97d-6e99-4d73-964a-62806b66fae0"
        edge_gateway_name               = "demo-gw"
      ~ fw_default_rule_action          = "deny" -> "accept"
        fw_default_rule_logging_enabled = true
        fw_enabled                      = true
        id                              = "urn:vcloud:gateway:68a6a97d-6e99-4d73-964a-62806b66fae0"
      ~ lb_acceleration_enabled         = false -> true
      ~ lb_enabled                      = true -> false
      ~ lb_logging_enabled              = false -> true
      ~ lb_loglevel                     = "info" -> "debug"
    }

Plan: 0 to add, 1 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

vcd_edgegateway_settings.egw-settings: Modifying... [id=urn:vcloud:gateway:68a6a97d-6e99-4d73-964a-62806b66fae0]
vcd_edgegateway_settings.egw-settings: Still modifying... [id=urn:vcloud:gateway:68a6a97d-6e99-4d73-964a-62806b66fae0, 10s elapsed]
vcd_edgegateway_settings.egw-settings: Still modifying... [id=urn:vcloud:gateway:68a6a97d-6e99-4d73-964a-62806b66fae0, 20s elapsed]
vcd_edgegateway_settings.egw-settings: Modifications complete after 21s [id=urn:vcloud:gateway:68a6a97d-6e99-4d73-964a-62806b66fae0]

Apply complete! Resources: 0 added, 1 changed, 0 destroyed.
```

Reply:
p.s. It looks there may even be a bug here, because I tried another scenario.

1. Remove all parameters from EG HCL that are available and defined in global settings resource.
2. Run apply.
3. Terraform still complains and tries to apply **some** of them through EG (note, `lb_enabled` is **not** complaining and that's the way it should be for all):
```
 # vcd_edgegateway.demo_gw will be updated in-place
  ~ resource "vcd_edgegateway" "demo_gw" {
        advanced                        = true
        configuration                   = "compact"
        default_external_network_ip     = "10.150.160.116"
        default_gateway_network         = "demo-ext-net"
        description                     = "VMworld Demo EG"
        distributed_routing             = false
        external_network_ips            = [
            "10.150.160.116",
        ]
        fips_mode_enabled               = false
      ~ fw_default_rule_action          = "accept" -> "deny"
      ~ fw_default_rule_logging_enabled = true -> false
        fw_enabled                      = true
        ha_enabled                      = false
        id                              = "urn:vcloud:gateway:68a6a97d-6e99-4d73-964a-62806b66fae0"
      ~ lb_acceleration_enabled         = true -> false
        lb_enabled                      = false
      ~ lb_logging_enabled              = true -> false
      ~ lb_loglevel                     = "debug" -> "info"
```

Reply:
Quoting from the documentation:

> The main use case of this resource is to allow both providers and tenants to change edge gateways global settings (such as enabling load balancing or firewall) when the edge gateway was created outside of terraform.

Given that, you can find many cases where you can abuse this resource if you use outside this work frame. 
If you can create an edge gateway, you have no good reason to use this additional resource.

Moreover, all the nuances of using this resource are well defined in the docs.

Reply:
> Given that, you can find many cases where you can abuse this resource if you use it outside this work frame.

> Moreover, all the nuances of using this resource are well defined in the docs.

I don't agree with this.

1. Docs do not state that there's a risk of collision between the properties of these two **resources**. There's only a warning about the use of this resource twice.
2. Try the above bug report with both resources defined, even if you don't define the conflicting parameters in the EG.

Reply:
I want to reiterate that using both an edge gateway resource (i.e. creating an edge gateway with Terraform) and a `vcd_edgegateway_settings` is unrealistic and not useful at all, given the destination of this resource, and given that you can change all attributes if you can create the resource in the first place. The use case is: 

* a data source for an edge gateway
* an instance of `vcd_edgegateway_settings` 

Moreover, there is information in the docs about what happens when using data source + `vcd_edgegateway_settings`:

> -> **Tip:** Although this resource changes values in the edge gateway referenced as a data source, due to how Terraform works, the state of the edge gateway doesn't get updated. To reconcile the state of the data source with the values modified in `vcd_edgegateway_settings`, you need to run `terraform refresh` after `apply`.

Reply:
> I want to reiterate that using both an edge gateway resource (i.e. creating an edge gateway with Terraform) and a vcd_edgegateway_settings is unrealistic and not useful at all

I see. Then, given that using both doesn't work (as demostrated above), it would be best if we didn't allow to create a `vcd_edgegateway_settings` that is pointing to a `vcd_edgegateway` resource in the same HCL. Would that be possible to do?
If not, then we need another warning in the docs clearly stating your above explanation.

Reply:
Tried the following, more practical, scenario, all with Terraform:
1. Create EG as provider **without** the global settings parameters.
2. Create global settings as tenant.
3. Run `terraform plan` as provider.

**Expected**

That the plan won't show differences.

**Result**

Plan shows diffs for at least the following fields, because it tries to re-set their values to defaults:

```
      ~ fw_default_rule_action          = "accept" -> "deny"
      ~ fw_default_rule_logging_enabled = true -> false
        fw_enabled                      = true
        ha_enabled                      = false
        id                              = "urn:vcloud:gateway:68a6a97d-6e99-4d73-964a-62806b66fae0"
      ~ lb_acceleration_enabled         = true -> false
```

Reply:
I don't see how you would expect no changes in the state after explicitly changing properties. 
Changing the properties with `vcd_edgegateway_settings` is the same as changing through the UI. After that, you would need to reconcile the state manually.
In this case, however, either operation n. 1 or n. 2 should be happening. 
There are two scenarios, as mentioned in the docs: 

* **Main scenario is to solve a problem when the entity was created without Terraform** and we want to allow tenants to change the settings.
* Secondary scenario is when the **providers create the resource and then delegate the changes to the tenant**. In this case, they **MUST** expect changes in the state, as this is what they told the tenants to do.

Reply:
To summarize, the main issue that those conflicting parameters have default values and there doesn't seem to be an elegant way of resolving that apart from removing them from EG and always using two resources as a bundle, which is not convenient.